### PR TITLE
ENH: Add axis parameter to add_prefix and add_suffix

### LINKED ISSUE #47819: ENH: Add `axis` parameter to `add_prefix` and `add_suffix`
### Feature Type

- [ ] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

`df.add_prefix` and `df.add_suffix` can be used on columns only.  
Would be super handy if it can be used on the index to.
Preferably with the option `inplace`.

Example:

```python
import pandas as pd
import numpy as np

np.random.seed(123)
data = np.random.randint(0, 100, size=(4, 3))
df = pd.DataFrame(data, index=range(1, 5), columns=range(20, 23))
df.index.name = 'Quarter'
```
DataFrame:
```
         20  21  22
Quarter            
1        66  92  98
2        17  83  57
3        86  97  96
4        47  73  32
```

New feature:
```python
df.add_prefix('Q', axis=0, inplace=True)
df.add_prefix(20, axis=1, inplace=True)
```

Expected result:
```
         2020  2021  2022
Quarter                  
Q1         66    92    98
Q2         17    83    57
Q3         86    97    96
Q4         47    73    32
```

Implementation suggestion:
```python
# columns
if df.columns is RangeIndex:
    df.columns = df.columns + 2000 # df.columns stays RangeIndex
else:
   df = df.add_prefix(20) # df.columns becomes list of integers

# index
if df.index is RangeIndex:
    df.set_index(df.index + 2000, inplace=True)
if all(isinstance(idx, int) for idx in list(df.index)): # df.index is list of integers
    df.set_index([int(str(prefix) + str(idx)) for idx in df.index], inplace=True)
else:
    df.set_index([str(prefix) + str(idx) for idx in df.index], inplace=True)
```

### Feature Description

```
df.add_prefix('Q', axis=0, inplace=True)
df.add_prefix(20, axis=1, inplace=True)
```

### Alternative Solutions

Implementation suggestion:
```python
# columns
if df.columns is RangeIndex:
    df.columns = df.columns + 2000 # df.columns stays RangeIndex
else:
   df = df.add_prefix(20) # df.columns becomes list of integers

# index
if df.index is RangeIndex:
    df.set_index(df.index + 2000, inplace=True)
if all(isinstance(idx, int) for idx in list(df.index)): # df.index is list of integers
    df.set_index([int(str(prefix) + str(idx)) for idx in df.index], inplace=True)
else:
    df.set_index([str(prefix) + str(idx) for idx in df.index], inplace=True)
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dannyi96**
- [X] closes #47819
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by dannyi96** [comment]
Hi @mroeschke, are these changes fine or should we hold changes in this API for now ?
Accordingly (if required) can update the whatsnew file
Thanks in advance. :)

**Review by phofl — COMMENTED** [review]
lgtm, but we need a whatsnew

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -334,7 +334,7 @@ Other enhancements
 - :meth:`Series.add_suffix`, :meth:`DataFrame.add_suffix`, :meth:`Series.add_prefix` and :meth:`DataFrame.add_prefix` support a ``copy`` argument. If ``False``, the underlying data is not copied in the returned object (:issue:`47934`)
 - :meth:`DataFrame.set_index` now supports a ``copy`` keyword. If ``False``, the underlying data is not copied when a new :class:`DataFrame` is returned (:issue:`48043`)
 - The method :meth:`.ExtensionArray.factorize` accepts ``use_na_sentinel=False`` for determining how null values are to be treated (:issue:`46601`)
-
+- :meth:`Series.add_suffix`, :meth:`DataFrame.add_suffix`, :meth:`Series.add_prefix` and :meth:`DataFrame.add_prefix` support an ``axis`` argument. If ``axis`` is unset, rows are considered for ``Series`` and columns for ``DataFrame``. If ``axis`` is set, this default behaviour can be overwritten (:issue:`48085`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.5.0.rst:337]: The new behaviour is consistent with the current behaviour?

Comment (dannyi96) [reply doc/source/whatsnew/v1.5.0.rst:337]: yup you're right. it's basically backward compatible.
- if user doesn't specify this ``axis`` parameter, it resorts to current behavior where rows are considered for ``Series`` and columns for ``DataFrame``.
- If user specifies this `axis` parameter, then the new behavior takes effect & add_prefix/add_suffix is done based on the axis provided.

Comment (phofl) [reply doc/source/whatsnew/v1.5.0.rst:337]: Then you can trim this a bit and just write that the default behavior can be overwritten without specifying what the current behavior is

Comment (dannyi96) [reply doc/source/whatsnew/v1.5.0.rst:337]: thanks @phofl for the suggestion. Makes sense.
will update this.

Comment (dannyi96) [reply doc/source/whatsnew/v1.5.0.rst]: Updated the whatsnew as per suggestion

**Review by phofl — APPROVED** [review]

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -334,6 +334,7 @@ Other enhancements
 - :meth:`Series.add_suffix`, :meth:`DataFrame.add_suffix`, :meth:`Series.add_prefix` and :meth:`DataFrame.add_prefix` support a ``copy`` argument. If ``False``, the underlying data is not copied in the returned object (:issue:`47934`)
 - :meth:`DataFrame.set_index` now supports a ``copy`` keyword. If ``False``, the underlying data is not copied when a new :class:`DataFrame` is returned (:issue:`48043`)
 - The method :meth:`.ExtensionArray.factorize` accepts ``use_na_sentinel=False`` for determining how null values are to be treated (:issue:`46601`)
+- :meth:`Series.add_suffix`, :meth:`DataFrame.add_suffix`, :meth:`Series.add_prefix` and :meth:`DataFrame.add_prefix` support an ``axis`` argument. If ``axis`` is set, the default behaviour of which axis to consider can be overwritten (:issue:`48085`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.5.0.rst:337]: Please link to the issue not the pr

Comment (dannyi96) [reply doc/source/whatsnew/v1.5.0.rst]: thanks, updated

File: pandas/core/generic.py [hunk 3]
```diff
@@ -4619,7 +4619,9 @@ def _update_inplace(self, result, verify_is_copy: bool_t = True) -> None:
         self._maybe_update_cacher(verify_is_copy=verify_is_copy, inplace=True)
 
     @final
-    def add_prefix(self: NDFrameT, prefix: str) -> NDFrameT:
```
Comment (dannyi96) [inline pandas/core/generic.py:4622]: has the copy parameter change which was previously present been removed ?

Comment (dannyi96) [reply pandas/core/generic.py:4622]: seems like it was done has part of #48201 
@mroeschke @phofl - can i ignore the above removal & add the axis parameter on top of what's now in main branch ?

Comment (dannyi96) [reply pandas/core/generic.py:4622]: have updated the PR with the new param changes on top of latest main branch

**Review by phofl — APPROVED** [review]
cc @mroeschke I think this is ready. If we don't backport, have to change versionadded

**Comment by mroeschke** [comment]
IMO I think this is more appropriate for 1.6, so I would be in favor of moving the release notes to `1.6.0.rst`.

**Comment by phofl** [comment]
No preference on my side. @dannyi96 could you move whatsnew and change Versionadded?

**Comment by dannyi96** [comment]
Sure thing @mroeschke @phofl. Will move the notes to 1.6

**Comment by dannyi96** [comment]
updated the whatsnew. 
( have` added it as part of "Other enhancements" in whatsnew v1.6 file )

File: pandas/tests/frame/methods/test_add_prefix_suffix.py [hunk 4]
```diff
@@ -18,3 +18,22 @@ def test_add_prefix_suffix(float_frame):
     with_pct_suffix = float_frame.add_suffix("%")
     expected = Index([f"{c}%" for c in float_frame.columns])
     tm.assert_index_equal(with_pct_suffix.columns, expected)
+
+
+def test_add_prefix_suffix_axis(float_frame):
```
Comment (mroeschke) [inline pandas/tests/frame/methods/test_add_prefix_suffix.py:23]: Could you add a similar test for `Series` as well (may need to make a new file in `pandas/tests/series/methods/`?

Comment (phofl) [reply pandas/tests/frame/methods/test_add_prefix_suffix.py:25]: I think we should be able to parametrize with ``frame_or_series``?

Comment (mroeschke) [reply pandas/tests/frame/methods/test_add_prefix_suffix.py:25]: Yeah, but it appears the file structure was intended to separate `test/frame/method` from `test/series/methods`? Also `axis=1` shouldn't be applicable for `Series`

Comment (dannyi96) [reply pandas/tests/frame/methods/test_add_prefix_suffix.py:25]: thanks @mroeschke @phofl for the suggestions.
Have added the testcases for Series in a new file under `test/series/methods`
Let me know if any changes / alternatives to try. TIA

File: pandas/core/generic.py [hunk 5]
```diff
@@ -4694,6 +4702,10 @@ def add_suffix(self: NDFrameT, suffix: str) -> NDFrameT:
         ----------
         suffix : str
             The string to add after each label.
+        axis : {{0 or 'index', 1 or 'columns', None}}, default None
+            Axis to add suffix on
+
+             .. versionadded:: 1.5.0
```
Comment (mroeschke) [inline pandas/core/generic.py:4708]:
```suggestion
             .. versionadded:: 1.6.0
```

**Review by mroeschke — COMMENTED** [review]
One small comment otherwise LGTM. If you could merge in main once more, that would be great.

File: pandas/core/generic.py (hunk 5 above)
Comment (dannyi96) [reply pandas/core/generic.py]: thanks 😄, had missed this. 
updated this in both the changed methods docstring.

**Comment by dannyi96** [comment]
> One small comment otherwise LGTM. If you could merge in main once more, that would be great.

done 🙂

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @dannyi96, great work!

**Pull request merged**